### ISSUE TITLE
Add moveDef parameters to control descent speed

### ISSUE THREAD

**Issue by Krogoth100** [beyond-all-reason/spring#392]
MoveDef's slow down on climbing is defined with **slopeMod** parameter.
This does not affect the descent speed however: descending goes seemingly as fast as with no slope at all, only climbing slows the unit. Which is completely fine for a wide range of cases.
But consider this: we want to make a unit with very good climbing ability. Not like all-terrain, but still quite good. Now, if you make it run downhill, it would look completely unrealistic/out of touch. Unit is supposed to step down carefully, not run at full speed ahead, so the slow must be applied like for climbing. But there should also be a minimum slope value, from which this slow starts to kick in. Because, naturally, unit should not be slowed at all if the slope change is small, so the effect does not become aggravating.
So I suggest to add 2 parameters: **descentSlopeMod** (default: 0) and **descentSlowSlope** (default: 0).
First one will give freedom to define a different slopeMod on descent. It could be same as slopeMod, of course, but I think such option is necessary to define good behaviour.
And the second parameter would define aforementioned slope value, from which the slow effect smoothly starts to kick in.

**Comment by marcushutchings** [comment]
Have you tried using this in your modrule settings?

`movement.allowDirectionalPathing = 0`

This option: `determines if ground speed going downhill != going uphill`

the default is true (1)

**Comment by Krogoth100** [comment]
> Have you tried using this in your modrule settings?
> 
> `movement.allowDirectionalPathing = 0`
> 
> This option: `determines if ground speed going downhill != going uphill`
> 
> the default is true (1)

Yes, but that's not it. A higher speed downhill is a correct behaviour.
The difference is by _how much_ it is quicker than a slowdown, that is important.

Correct me if I am wrong, because I am not a Spring mapmaker, but I have got an impression that the authors purposefully put speed modifiers for bots on cliffs, slowing them, to achieve what somehow resembles an intended ascension/descension effect. Which by itself may suggest that the current system is lacking somewhat (and that is not being fixed via existing moveDef properties).
That's an observation I got, but it may not be entirely correct, because I am not a mapmaker.

But when I tried to make a map of my own, without any speed mod "tricks", I found that my units behaving rather bad. Basically, if you get a combination of a good movement speed (Zipper-tier), climbing ability and quite a slope, when they descend - that's gonna look bad. Units don't just run like berserk, it looks like they are just crashing down without any safety. : S

What should be the fix, as I see it, is that _descension slowdown_ accumulates much _slower_ than ascension, but _still reaches about the same value_ when the slope becomes steep. The change in move speed will be more abrupt. It should look like bots lose their confidence in ground and step carefully. It can be further accentuated with a good special walk animation for that kind of a slowdown.

**Comment by Krogoth100** [comment]
Yeah, I am gonna clarify that "my observation" better.
The suspicion is that mapmakers needed to reduce _descension speed_ particularly. That's why they add a thing like a 50% speed mod wherever a slope is quite steep.
My assumption is, with an advanced formula they would be able to modify moveDefs instead and clear their maps of the most speed mod interventions.

**Comment by marcushutchings** [comment]
@Beherith do you the answer to the questions about map making and influencing units speeds on slopes, above?

**Comment by Beherith** [comment]
Speed modifiers are almost never placed on slopes, except old dsd. But the typemap there should be fully ignored anyway. They are used to define difficult/impassable terrain and fast roads. Adding the time domain as you suggest to descent and ascent would be very difficult for pathfinding. Use set wanted max speed in a gadget instead imo.

**Comment by Beherith** [comment]
> > Have you tried using this in your modrule settings?
> > 
> > `movement.allowDirectionalPathing = 0`
> > 
> > This option: `determines if ground speed going downhill != going uphill`
> > 
> > the default is true (1)
> 
> Yes, but that's not it. A higher speed downhill (compared to uphill) is a correct behaviour.
> The difference is by _how much_ and _when_ descension is quicker than ascension, that is important.
> 
> Correct me if I am wrong, because I am not a Spring mapmaker, but I have got an impression that the authors purposefully put speed modifiers for bots (just as for vehs) on cliffs, slowing them, to achieve what somehow resembles an intended ascension/descension effect. Which by itself may suggest that the current system is lacking somewhat (and that is not being fixed via existing moveDef properties).
> That's an observation I got, but it may not be entirely correct, because I am not a mapmaker.
> 

> What should be the fix, as I see it, is that _descension slowdown_ accumulates much _slower_ than ascension, but _still reaches about the same value_ when the slope becomes steep. The change in move speed will be more abrupt. It should look like bots lose their confidence in ground and step carefully. It can be further accentuated by a developer with a good special walk animation for that kind of a slowdown.

So, in general, if you want symmetric slope speed up and down, use typemap and slopemod. All bar bot animations are fully aware of the units current velocity and are played back accordingly. In fact, skeletor even supports not just dynamic animation speed, but dynamic animation amplitude.

**Comment by Krogoth100** [comment]
> Speed modifiers are almost never placed on slopes, except old dsd. But the typemap there should be fully ignored anyway. They are used to define difficult/impassable terrain and fast roads. Adding the time domain as you suggest to descent and ascent would be very difficult for pathfinding. Use set wanted max speed in a gadget instead imo.

Good to know! Looks like I overestimated the importance of speedmods.

Symmetric slope won't do the thing, it has to have different values on both ends to make it work.

I have made a short video, illustrating the problem, but bear in mind that in my game the problem is more severe:
https://youtu.be/yivCBAVRUkI
(FPS not too good, sorry for that. The map is Desert Triad, quite a good example of a steep slope)

Now I will go into some detail that my game has about landscape. In my game the rendered, visual landscape is decoupled from physical. What is rendered has much more resolution than even 16K resolution height map. There are some techniques, some tricks to do that. Overall it works, but the point is, that steep slopes have geometric detail that makes it look convincing when units tread carefully and not rush themselves down. For example, I could literally do staircases and make bots walk them (not without bad leg animations though, because I am noob in that department rn). It would be visually a sequence of horizontal and vertical slopes, but physically (engine-side height map): a slope of 45.
Unit's visual Y coordinate is updated according to on-screen landscape detail, so the closer is piece of landscape to the viewer, the more correction a unit gets. But what _must be_ precise, that is XZ coordinate. If certain Y decoupling is tolerable, like a missile missing unit's head when it shouldn't, then XZ imprecision is not. So physical XZ movement must be as good as possible. If I want to have a convincingly looking slope with roughs and rocks, I need a way to slowdown those units when they go downhill.
But if make it symmetrical? That would ruin it for small slopes. Exactly on small/medium slopes - that's when units' (especially bots') movement speed should look impressive. That is why some kind of a buffer/threshold to make it asymmetrical would solve the problem.

**Comment by Krogoth100** [comment]
Also, I just had this thought...
( This is just some secondary idea to accompany main one )
The behaviour, that units aren't slowed down on descension.. This is what we roughly use to imitate unit's jump, right? If we look at some Flea from afar, we don't really need to see if its height changes, its visually too small. Flea just looks like jumping when running from a smaller slope, and that is very well.
So does a race car like T1 Scout Vehicle. Completely normal behaviour. If anyone wanted, they could do some shader tricks and make it even visually jump - that's not to say, you should, it's just this behaviour feels that much right.
But what about some kind of a heavy tank, or a heavy construction vehicle? That's what I am leading to, that some thread chassis heavy units shouldn't have that effect and probably need to slowdown even on a smaller slope downhill. That's where aforementioned threshold finds another use. Putting it to nearly zero will ensure almost symmetrical behaviour, which _some units_ I think should have.
Again, this is not very important point, but a small detail I think worth noticing.

**Comment by Beherith** [comment]
Vid marked as private. I would still stick with typemap for this effect.

**Comment by Krogoth100** [comment]
> Vid marked as private. I would still stick with typemap for this effect.

Thanks for letting me know! May be geo-storm affecting brain activity.
Fixed
https://youtu.be/yivCBAVRUkI

**Comment by Krogoth100** [comment]
I am gonna use typemap as a last resort. But that doesn't seem a good way - tweaking, fixing what should already work. I mean, a move def must describe behavior well, that's what it partly is for - to describe interactions with slope. But because the feature is limited and doesn't do fully its job, have to process the map and apply adjustments.
